### PR TITLE
feat: add support for ghost mode to make a CAD model appear translucent

### PR DESCRIPTION
**Pull request by larsmoa**

### CODE REVIEW

File: viewer/src/glsl/base/updateFragmentColor.glsl [hunk 1]
```diff
@@ -53,7 +54,7 @@ void updateFragmentColor(int renderMode, vec4 color, float treeIndex, vec3 norma
     } else if (renderMode == RenderTypeDepth) {
         gl_FragColor = packDepthToRGBA(depth);
     } else {
-        gl_FragColor = vec4(1.0, 0.0, 0.0, 1.0);
+        gl_FragColor = vec4(1.0, 0.0, 1.0, 1.0);
```
Comment (christjt) [inline viewer/src/glsl/base/updateFragmentColor.glsl:57]: Is this supposed to be an error color? When does this even ever happen, and why purple?

File: viewer/src/public/migration/Cognite3DModel.ts [hunk 2]
```diff
@@ -521,7 +574,7 @@ export class Cognite3DModel extends THREE.Object3D implements CogniteModelBase {
     if (makeGray) {
       throw new NotSupportedInMigrationWrapperError('makeGray is not supported');
     }
-    for (let i = 0; i < this.cadModel.scene.maxTreeIndex; i++) {
+    for (let i = 0; i <= this.cadModel.scene.maxTreeIndex; i++) {
```
Comment (christjt) [inline viewer/src/public/migration/Cognite3DModel.ts:577]: Damn, nice catch!

File: viewer/src/datamodels/cad/rendering/EffectRenderManager.ts [hunk 3]
```diff
@@ -11,34 +11,39 @@ import { CadNode } from '..';
 import { Cognite3DModel } from '@/migration';
 import { RootSectorNode } from '../sector/RootSectorNode';
 
+type Object3DStructure = {
```
Comment (christjt) [inline viewer/src/datamodels/cad/rendering/EffectRenderManager.ts:14]: This class is starting to get a little messy, and I think it is ripe for refactoring everything that has to do with scene building into its own component / class, and let the render manager deal with rendering and and render targets. Not something we necessarily have to do right now, but I think we should at the very least track it under tech debt.

File: viewer/src/glsl/base/updateFragmentColor.glsl [hunk 4]
```diff
@@ -21,15 +22,15 @@ void updateFragmentColor(int renderMode, vec4 color, float treeIndex, vec3 norma
...
-
         vec2 cap = normal.xy * 0.5 + 0.5;
-
         vec4 mc = vec4(texture2D(matCapTexture, cap).rgb, 1.0);
         
         gl_FragColor = vec4(albedo.rgb * mc.rgb * 1.7, color.a);
-
+    } else if (renderMode == RenderTypeGhost) {
+        float amplitude = max(0.0, dot(normal, vec3(0.0, 0.0, 1.0)));
+        vec3 albedo = min(vec3(0.8, 0.8, 0.8) * (0.4 + 0.6 * amplitude), 1.0);
```
Comment (christjt) [inline viewer/src/glsl/base/updateFragmentColor.glsl:32]: NIT: vec3(0.8) == vec3(0.8, 0.8, 0.8)

**Review by christjt — CHANGES_REQUESTED** [review]

File: examples/src/pages/Testable.tsx [hunk 5]
```diff
@@ -37,13 +37,19 @@ export function Testable() {
...
+          let style = reveal.DefaultNodeAppearance.NoOverrides;
           if (pickedNodes.has(treeIndex)) {
-            return reveal.DefaultNodeAppearance.Highlighted;
+            style = {...style, ...reveal.DefaultNodeAppearance.Highlighted };
           }
-          return reveal.DefaultNodeAppearance.NoOverrides;
+          if (ghostedNodes.has(treeIndex)) {
+            style = {...style, ...reveal.DefaultNodeAppearance.Ghosted };
+          }
+          console.log(treeIndex, style);
```
[the remark this answers is not in the document]
Comment (larsmoa) [reply examples/src/pages/Testable.tsx:51]: Protip: use suggestions-feature in comments to make it super-easy to do these types of changes :)

File: examples/src/pages/Testable.tsx [hunk 6]
```diff
@@ -112,6 +118,18 @@ export function Testable() {
...
+        camera = new THREE.PerspectiveCamera();
+        position.x = 45;
+        position.y = 3;
+        position.z = 13;
+
+        const ghostedTreeIndices = [...Array(32).keys()].map((_value, index) => 2*index);
+        ghostedTreeIndices.forEach(p => ghostedNodes.add(p));
+        const highlightedTreeIndexes = [0,1,2,3,4,5,6,20,22,33,34,35];
+        highlightedTreeIndexes.forEach(p => pickedNodes.add(p));
+        console.log(ghostedTreeIndices);
```
Comment (larsmoa) [inline examples/src/pages/Testable.tsx:131]:
```suggestion
```

File: viewer/src/glsl/base/updateFragmentColor.glsl (hunk 4 above)
Comment (larsmoa) [reply viewer/src/glsl/base/updateFragmentColor.glsl:32]: oooh, didn't know actually

Comment (larsmoa) [inline viewer/src/glsl/base/updateFragmentColor.glsl:32]:
```suggestion
        vec3 albedo = min(vec3(0.8) * (0.4 + 0.6 * amplitude), 1.0);
```

File: viewer/src/glsl/base/updateFragmentColor.glsl (hunk 1 above)
Comment (larsmoa) [inline viewer/src/glsl/base/updateFragmentColor.glsl:57]:
```suggestion
        // Unknown render mode - should not happen. 
        gl_FragColor = vec4(1.0, 0.0, 1.0, 1.0);
```

File: viewer/src/datamodels/cad/sector/culling/ByVisibilityGpuSectorCuller.ts [hunk 7]
```diff
@@ -120,7 +120,7 @@ class TakenSectorMap {
  * and loads sectors based on priority within a budget.
  */
 export class ByVisibilityGpuSectorCuller implements SectorCuller {
-  public static readonly DefaultCostLimit = 50 * 1024 * 1024;
+  public static readonly DefaultCostLimit = 35 * 1024 * 1024;
```
[the remark this answers is not in the document]
Comment (larsmoa) [reply viewer/src/datamodels/cad/sector/culling/ByVisibilityGpuSectorCuller.ts:123]: Oh, definitely. Unintentional, nice catch

Comment (larsmoa) [inline viewer/src/datamodels/cad/sector/culling/ByVisibilityGpuSectorCuller.ts:123]:
```suggestion
  public static readonly DefaultCostLimit = 50 * 1024 * 1024;
```

File: examples/src/pages/Testable.tsx (hunk 5 above)
Comment (larsmoa) [inline examples/src/pages/Testable.tsx:51]:
```suggestion
```

File: viewer/src/glsl/base/updateFragmentColor.glsl (hunk 1 above)
Comment (larsmoa) [reply viewer/src/glsl/base/updateFragmentColor.glsl:58]: Just a safe guard should we introduce a new RenderMode and forget to update check (or provide wrong value somehow).

File: viewer/src/datamodels/cad/rendering/EffectRenderManager.ts (hunk 3 above)
Comment (larsmoa) [reply viewer/src/datamodels/cad/rendering/EffectRenderManager.ts]: Refactored this, have another looksie and see what you think

**Review by christjt — APPROVED** [review]

**Pull request merged**